# World Bank Development Indicators Analysis

## Notebook 01: Data Import & Exploration

### Objective
The purpose of this notebook is to understand the structure of the World Bank Development Indicators (WDI) dataset before performing any data cleaning or analysis.

### Tasks
- Import the libraries & dataset
- Explore workbook structure
- Understand each worksheet
- Examine dataset dimensions
- Identify data types
- Detect potential data quality issues

In [1]:
import pandas as pd

In [3]:
import numpy as np

In [6]:
excel_file="../data/raw/WDIEXCEL.xlsx"
type(excel_file)

str

# Explore the Excel Sheet

In [9]:
xls=pd.ExcelFile(excel_file)
type(xls)
xls.sheet_names

['Data', 'Country', 'Series', 'country-series', 'series-time', 'footnote']

In [10]:
df_data = pd.read_excel(excel_file, sheet_name="Data")

In [11]:
df_data.shape

(396970, 70)

In [12]:
df_data.columns

Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021', '2022',
       '2023', '2024', '2025'],
      dtype='str')

In [13]:
df_data.columns.to_list()

['Country Name',
 'Country Code',
 'Indicator Name',
 'Indicator Code',
 '1960',
 '1961',
 '1962',
 '1963',
 '1964',
 '1965',
 '1966',
 '1967',
 '1968',
 '1969',
 '1970',
 '1971',
 '1972',
 '1973',
 '1974',
 '1975',
 '1976',
 '1977',
 '1978',
 '1979',
 '1980',
 '1981',
 '1982',
 '1983',
 '1984',
 '1985',
 '1986',
 '1987',
 '1988',
 '1989',
 '1990',
 '1991',
 '1992',
 '1993',
 '1994',
 '1995',
 '1996',
 '1997',
 '1998',
 '1999',
 '2000',
 '2001',
 '2002',
 '2003',
 '2004',
 '2005',
 '2006',
 '2007',
 '2008',
 '2009',
 '2010',
 '2011',
 '2012',
 '2013',
 '2014',
 '2015',
 '2016',
 '2017',
 '2018',
 '2019',
 '2020',
 '2021',
 '2022',
 '2023',
 '2024',
 '2025']

In [14]:
df_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 396970 entries, 0 to 396969
Data columns (total 70 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   Country Name    396970 non-null  str    
 1   Country Code    396970 non-null  str    
 2   Indicator Name  396970 non-null  str    
 3   Indicator Code  396970 non-null  str    
 4   1960            37226 non-null   float64
 5   1961            42560 non-null   float64
 6   1962            43847 non-null   float64
 7   1963            44777 non-null   float64
 8   1964            45283 non-null   float64
 9   1965            47096 non-null   float64
 10  1966            47354 non-null   float64
 11  1967            47991 non-null   float64
 12  1968            48608 non-null   float64
 13  1969            49391 non-null   float64
 14  1970            72137 non-null   float64
 15  1971            76468 non-null   float64
 16  1972            78158 non-null   float64
 17  1973            78812

In [15]:
df_data.head()

,Country Name,Country Code,Indicator Name,Indicator Code,1960,1961,1962,1963,1964,1965,...,2016,2017,2018,2019,2020,2021,2022,2023,2024,2025
0,Africa Eastern and Southern,AFE,Access to clean fuels and technologies for coo...,EG.CFT.ACCS.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,18.685118,19.205632,19.742772,20.332679,20.862800,21.419621,21.996456,22.541440,NaN,NaN
1,Africa Eastern and Southern,AFE,Access to clean fuels and technologies for coo...,EG.CFT.ACCS.RU.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,7.606712,7.926604,8.309896,8.704591,9.106640,9.480804,9.903209,10.288154,NaN,NaN
2,Africa Eastern and Southern,AFE,Access to clean fuels and technologies for coo...,EG.CFT.ACCS.UR.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,39.052626,39.321068,39.649534,39.968299,40.354628,40.723805,41.026351,41.289974,NaN,NaN
3,Africa Eastern and Southern,AFE,Access to electricity (% of population),EG.ELC.ACCS.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,38.859598,40.223744,43.035073,44.390861,46.282371,48.127211,48.742043,50.667516,52.598171,NaN
4,Africa Eastern and Southern,AFE,"Access to electricity, rural (% of rural popul...",EG.ELC.ACCS.RU.ZS,NaN,NaN,NaN,NaN,NaN,NaN,...,24.584906,25.375037,26.941984,28.983183,30.909991,32.709837,33.998179,35.240236,36.080572,NaN


In [17]:
df_data["Country Name"].unique()
df_data["Country Name"].nunique()

265

In [19]:
df_data["Indicator Name"].unique()
df_data["Indicator Name"].nunique()

1498

In [21]:
df_data[["Country Name","Indicator Name"]].head()

,Country Name,Indicator Name
0,Africa Eastern and Southern,Access to clean fuels and technologies for coo...
1,Africa Eastern and Southern,Access to clean fuels and technologies for coo...
2,Africa Eastern and Southern,Access to clean fuels and technologies for coo...
3,Africa Eastern and Southern,Access to electricity (% of population)
4,Africa Eastern and Southern,"Access to electricity, rural (% of rural popul..."


In [25]:
df_data.duplicated()
df_data.duplicated().sum()

np.int64(0)

In [27]:
df_data.isnull().sum()

Country Name           0
Country Code           0
Indicator Name         0
Indicator Code         0
1960              359744
                   ...  
2021              185497
2022              198974
2023              212798
2024              246812
2025              315299
Length: 70, dtype: int64

In [33]:
print(f"Rows: {df_data.shape[0]}")
print(f"Columns: {df_data.shape[1]}")
print(f"Countries: {df_data["Country Name"].nunique()}")
print(f"Indicators: {df_data["Indicator Name"].nunique()}")
print(f"Duplicate Rows: {df_data.duplicated().sum()}")

Rows: 396970
Columns: 70
Countries: 265
Indicators: 1498
Duplicate Rows: 0


In [45]:
df_data["Indicator Name"].value_counts()

Indicator Name
Access to clean fuels and technologies for cooking (% of population)                              265
Access to clean fuels and technologies for cooking, rural (% of rural population)                 265
Access to clean fuels and technologies for cooking, urban (% of urban population)                 265
Access to electricity (% of population)                                                           265
Access to electricity, rural (% of rural population)                                              265
                                                                                                 ... 
Women who believe a husband is justified in beating his wife when she refuses sex with him (%)    265
Women who were first married by age 15 (% of women ages 20-24)                                    265
Women who were first married by age 18 (% of women ages 20-24)                                    265
Women's share of population ages 15+ living with HIV (%)           

# Observations:

* The workbook has 6 sheets
* Dataset is in wide format
* Dataset contains 3,96,970 rows & 70 columns
* The first 4 columns are of data type "string" and the rest are all "float" values
* Large number of missing values exist but 0 duplicate rows
* Total Unique Countries: 265
* Total Unique Indicators: 1498

## Explore Remaining Subsheets

In [34]:
df_country = pd.read_excel(excel_file, sheet_name="Country")

In [35]:
df_country.shape

(264, 31)

In [37]:
df_country.columns.to_list()

['Country Code',
 'Short Name',
 'Table Name',
 'Long Name',
 '2-alpha code',
 'Currency Unit',
 'Special Notes',
 'Region',
 'Income Group',
 'WB-2 code',
 'National accounts base year',
 'National accounts reference year',
 'SNA price valuation',
 'Lending category',
 'Other groups',
 'System of National Accounts',
 'Alternative conversion factor',
 'PPP survey year',
 'Balance of Payments Manual in use',
 'External debt Reporting status',
 'System of trade',
 'Government Accounting concept',
 'IMF data dissemination standard',
 'Latest population census',
 'Latest household survey',
 'Source of most recent Income and expenditure data',
 'Vital registration complete',
 'Latest agricultural census',
 'Latest industrial data',
 'Latest trade data',
 'Latest water withdrawal data']

In [38]:
df_country.info()

<class 'pandas.DataFrame'>
RangeIndex: 264 entries, 0 to 263
Data columns (total 31 columns):
 #   Column                                             Non-Null Count  Dtype  
---  ------                                             --------------  -----  
 0   Country Code                                       264 non-null    str    
 1   Short Name                                         264 non-null    str    
 2   Table Name                                         264 non-null    str    
 3   Long Name                                          264 non-null    str    
 4   2-alpha code                                       262 non-null    str    
 5   Currency Unit                                      217 non-null    str    
 6   Special Notes                                      134 non-null    str    
 7   Region                                             217 non-null    str    
 8   Income Group                                       217 non-null    str    
 9   WB-2 code            

In [39]:
df_country.head()

,Country Code,Short Name,Table Name,Long Name,2-alpha code,Currency Unit,Special Notes,Region,Income Group,WB-2 code,...,Government Accounting concept,IMF data dissemination standard,Latest population census,Latest household survey,Source of most recent Income and expenditure data,Vital registration complete,Latest agricultural census,Latest industrial data,Latest trade data,Latest water withdrawal data
0,ABW,Aruba,Aruba,Aruba,AW,Aruban florin,NaN,Latin America & Caribbean,High income,AW,...,NaN,Enhanced General Data Dissemination System (e-...,2020 (expected),NaN,NaN,Yes,NaN,NaN,2018.0,NaN
1,AFE,Africa Eastern and Southern,Africa Eastern and Southern,Africa Eastern and Southern,ZH,NaN,"26 countries, stretching from the Red Sea in t...",NaN,NaN,ZH,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,AFG,Afghanistan,Afghanistan,Islamic State of Afghanistan,AF,Afghan afghani,The reporting period for national accounts dat...,Middle East & North Africa,Low income,AF,...,Consolidated central government,Enhanced General Data Dissemination System (e-...,1979,Multiple Indicator Cluster Survey 2022-2023,"Integrated household survey (IHS), 2016/17",NaN,2003.0,NaN,2018.0,NaN
3,AFW,Africa Western and Central,Africa Western and Central,Africa Western and Central,ZI,NaN,"22 countries, stretching from the westernmost ...",NaN,NaN,ZI,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,AGO,Angola,Angola,People's Republic of Angola,AO,Angolan kwanza,The World Bank systematically assesses the app...,Sub-Saharan Africa,Lower middle income,AO,...,Budgetary central government,Enhanced General Data Dissemination System (e-...,2014,Inquérito de Indicadores Múltiplos e de Saúde ...,"Integrated household survey (IHS), 2008/09",NaN,2023.0,NaN,2018.0,NaN


In [43]:
print(f"Unique Regions: {df_country["Region"].nunique()}")
print(f"How many countries in each region: {df_country["Region"].value_counts()}")

Unique Regions: 7
How many countries in each region: Region
Europe & Central Asia         58
Sub-Saharan Africa            48
Latin America & Caribbean     42
East Asia & Pacific           37
Middle East & North Africa    23
South Asia                     6
North America                  3
Name: count, dtype: int64


In [46]:
df_series = pd.read_excel(excel_file, sheet_name="Series")

In [47]:
df_series.shape

(1498, 20)

In [48]:
df_series.columns.to_list()

['Series Code',
 'Topic',
 'Indicator Name',
 'Short definition',
 'Long definition',
 'Unit of measure',
 'Periodicity',
 'Base Period',
 'Other notes',
 'Aggregation method',
 'Limitations and exceptions',
 'Notes from original source',
 'General comments',
 'Source',
 'Statistical concept and methodology',
 'Development relevance',
 'Related source links',
 'Other web links',
 'Related indicators',
 'License Type']

In [49]:
df_series.info()

<class 'pandas.DataFrame'>
RangeIndex: 1498 entries, 0 to 1497
Data columns (total 20 columns):
 #   Column                               Non-Null Count  Dtype  
---  ------                               --------------  -----  
 0   Series Code                          1498 non-null   str    
 1   Topic                                1498 non-null   str    
 2   Indicator Name                       1498 non-null   str    
 3   Short definition                     0 non-null      float64
 4   Long definition                      1498 non-null   str    
 5   Unit of measure                      1196 non-null   str    
 6   Periodicity                          1467 non-null   str    
 7   Base Period                          0 non-null      float64
 8   Other notes                          461 non-null    str    
 9   Aggregation method                   1128 non-null   str    
 10  Limitations and exceptions           892 non-null    str    
 11  Notes from original source           0 no

In [50]:
df_series.head()

,Series Code,Topic,Indicator Name,Short definition,Long definition,Unit of measure,Periodicity,Base Period,Other notes,Aggregation method,Limitations and exceptions,Notes from original source,General comments,Source,Statistical concept and methodology,Development relevance,Related source links,Other web links,Related indicators,License Type
0,AG.CON.FERT.PT.ZS,Environment: Agricultural production,Fertilizer consumption (% of fertilizer produc...,NaN,Fertilizer consumption measures the quantity o...,% (ratio),Annual,NaN,The world and regional aggregate series do not...,Weighted average,The FAO has revised the time series for fertil...,NaN,NaN,"FAO electronic files and web site, Food and Ag...",Methodology: Fertilizer consumption measures t...,"Factors, such as the green revolution, have le...",NaN,NaN,NaN,CC BY-4.0
1,AG.CON.FERT.ZS,Environment: Agricultural production,Fertilizer consumption (kilograms per hectare ...,NaN,Fertilizer consumption measures the quantity o...,kg per hectare of arable land,Annual,NaN,The world and regional aggregate series do not...,Weighted average,The FAO has revised the time series for fertil...,NaN,NaN,"FAO electronic files and web site, Food and Ag...",Methodology: Fertilizer consumption measures t...,"Factors, such as the green revolution, have le...",NaN,NaN,NaN,CC BY-4.0
2,AG.LND.AGRI.K2,Environment: Land use,Agricultural land (sq. km),NaN,Agricultural land refers to the land area that...,square kilometers (sq. km),Annual,NaN,Areas of former states are included in the suc...,Sum,The data are collected by the Food and Agricul...,NaN,NaN,"FAO electronic files and web site, Food and Ag...",Methodology: Agricultural land constitutes onl...,Agricultural land covers more than one-third o...,NaN,NaN,NaN,CC BY-4.0
3,AG.LND.AGRI.ZS,Environment: Land use,Agricultural land (% of land area),NaN,Agricultural land refers to the share of land ...,% (share) of land area,Annual,NaN,Areas of former states are included in the suc...,Weighted average,The data are collected by the Food and Agricul...,NaN,NaN,"FAO electronic files and web site, Food and Ag...",Methodology: Agriculture is still a major sect...,Agricultural land covers more than one-third o...,NaN,NaN,NaN,CC BY-4.0
4,AG.LND.ARBL.HA,Environment: Land use,Arable land (hectares),NaN,Arable land (in hectares) includes land define...,hectares,Annual,NaN,Areas of former states are included in the suc...,NaN,The Food and Agriculture Organization (FAO) tr...,NaN,NaN,"FAO electronic files and web site, Food and Ag...",Methodology: Temporary fallow land refers to l...,Agricultural land covers more than one-third o...,NaN,NaN,NaN,CC BY-4.0


In [53]:
df_series.duplicated().sum()

np.int64(0)

# Findings:

* Primary analytical worksheets are Data (Facts Table) - Main Dataset & Country / Series (Dimension Table / Metadata)
* Country sheet can be linked to the main Data using "Country Code"
* "Series Code" in series sheet is same as "Indicator Code" in data sheet
* We will need to convert the wide format data sheet to long format for simplifying SQL queries